[종근]  QnA 3단계(RESTful API) 구현 완료

Bellroute opened the pull request:
- ApiQuestionAcceptanceTest 구현
- ApiQuestionController 구현
- ApiAnswerAcceptanceTest 구현
- QnaServiceTest에 Answer에 대한 단위 테스트 구현
- QnaService에 Answer에 대한 기능 구현
- ApiAnswerController 구현
- Resource create, get 중복 제거

Bellroute committed:
refactor: AcceptancTest에 defaultQuestion() 반영

Bellroute committed:
feat: ApiQuestionAcceptanceTest

Bellroute committed:
feat: ApiQuestionController

Bellroute committed:
feat: AcceptanceTest에 defaultAnswer() 구현

Bellroute committed:
feat: ApiAnswerAcceptanceTest

Bellroute committed:
feat: Answer에 equalsContents(), delete() 구현

Bellroute committed:
feat: QnaServiceTest에answer에 대한 테스트 구현

Bellroute committed:
feat: QnaService에answer에 대한 기능 구현

Bellroute committed:
feat: ApiAnswerController 구현

Bellroute committed:
refactor: Resource create&get 중복 제거

wijehyeon commented on src/main/java/codesquad/domain/Answer.java:
```diff
@@ -76,4 +88,19 @@ public String generateUrl() {
+
+    public void delete(User loginUser) throws CannotDeleteException {
+        if (!isOwner(loginUser)) {
+            throw new CannotDeleteException("delete error");
```
cannot delete Exception은 무슨 이유로 cannot delete한지 자세히 알려주지 않는 예외같습니다. 
인증에러라는 걸 알려줄 수 있는 예외를 찾아 써봅시다

hyukjin-lee replied:
메시지로 적어주면 어디서 무슨 이유로 예외가 발생했는지 서버 로그에 찍힙니다.
그리고 controller advice에서 에러처리를 할건데 그곳에 CannotDeleteException 매핑이 되어있는걸 프론트에 httpstatus 내려줄때 unAuthorizedException을 내려주면 됩니다.

SimKiSeong commented on src/main/java/codesquad/domain/Answer.java:
```diff
@@ -1,12 +1,21 @@
+import java.util.Objects;
 
 @Entity
+@Getter
```
lombok 쓸때 getter setter 대신에 data쓰면 getter setter가 한번에 처리됩니다.

dadadamarine replied:
https://projectlombok.org/features/Data
@Data 어노테이션은 getter, setter들을 모든 필드에 생성하기때문에 주의해서 써야하기도 합니다.
또한 @RequiredArgsConstructor 어노테이션도 같이 적용합니다.

Bellroute replied:
넵 getter와 setter만 적용하고 싶어서 data를 사용하지 않았습니다

SimKiSeong commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -64,14 +67,36 @@ public boolean isQuestionWriter(User loginUser, long id) {
+    public Answer updateAnswer(User loginUser, long id, AnswerDTO answerDTO) {
+        Answer answer = answerRepository.findById(id).orElseThrow(NoSuchElementException::new);
+
+        if (!answer.isOwner(loginUser)) {
```
유저를 채크하고 update하는 과정을 도메인에서 하면 실수가 줄어든다고 합니다.

dadadamarine replied:
다른사람이 Answer라는 객체를 사용하게 될때 , 그사람이 따로 권한 체크를 하는걸 까먹었다고 하더라도, 도메인 레벨에서 권한체크를 해주니 문제가 생길 가능성이 줄어듭니다.

Bellroute replied:
그런 이점이 있군요. 반영하겠습니다

SimKiSeong commented on src/test/java/codesquad/web/ApiQuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,60 @@
+
+    @Test
+    public void create() {
+        ResponseEntity<Question> response = createQuestionResource("/api/questions", defaultQuestion());
```
/api/questions가 다른 메소드에서도 중복으로 사용되고 있습니다. 상수로 만들어서 관리하면 편할듯 합니다.

SimKiSeong commented on src/test/java/codesquad/web/ApiAnswerAcceptanceTest.java:
```diff
@@ -0,0 +1,86 @@
+        String location = String.format("/api/questions/%d/answers/%d", defaultQuestion().getId(), defaultAnswer().getId());
+
+        Answer updateAnswer = new Answer(defaultUser(), "content");
+        response = basicAuthTemplate(newUser("testuser1"))
```
test유저를 반복해서 쓰는데 하나로 만들어서 쓰는 것도 좋을 것 같습니다.

SimKiSeong reviewed (commented):
고생하셨습니다.

dadadamarine commented on src/test/java/codesquad/web/ApiAnswerAcceptanceTest.java:
```diff
@@ -0,0 +1,86 @@
+    @Test
+    public void show() {
+        ResponseEntity<Answer> response = createAnswerResource(String.format("/api/questions/%d/answers", defaultQuestion().getId()), defaultAnswer());
+        String location = String.format("/api/questions/%d", defaultQuestion().getId());
```
이부분은 궁금해서 여쭤봅니다. Answer의 create는 컨트롤러 상에 AnswerDto객체로 받아오게 되어있습니다. 
테스트에서 보면 requestBody에 Answer클래스를 보내는것 같은데 정상적으로 작동하나요?

작동한다면, 어떻게 Answer객체를 AnswerDto객체로 매핑해서 컨트롤러에서 받을수 있는건가요?

dadadamarine commented on src/main/java/codesquad/domain/Answer.java:
```diff
@@ -76,4 +88,19 @@ public String generateUrl() {
+            return false;
+        }
+
+        return this.contents.equals(target.contents);
```
데이터의 equals 판단은 id로 하는게 더 맞는것 같습니다.

Bellroute replied:
해당 메소드가 인수테스트 단에서 사용 되고 있는데요 테스트에서 업데이트용 answer객체를 새로 생성해버려서 기존 answer객체와 id값이 달라질 것을 우려해서 contents를 비교했습니다. 새로운 객체를 생성해서 업데이트하는 방식이 문제가 될까요?

hyukjin-lee replied:
테스트만을 위한 프로덕션 코드를 작성 했다는 것은 테스트를 잘못 짰다는 뜻입니다. 콘텐츠만 같으면 같은 게시물입니까? 고유값인 id나  객체 전체가 같아야죠. 테스트는 통과할진 몰라도 논리적으로 완벽하지 않은 테스트여서 언젠간 문제가 될 가능성이 있습니다. 그리고 테스트에서만 사용되는 코드인게 제일 큰 흠이네요

dadadamarine commented on src/main/java/codesquad/domain/AnswerRepository.java:
```diff
@@ -2,6 +2,8 @@
+import java.util.Optional;
 
+public interface AnswerRepository extends JpaRepository<Answer, Long> {
+    Optional<Answer> findAnswerByWriter(User user);
```
Answer은 적어도 좋지만, 관례상 생략하는 경우가 많으므로 생략해주는게 좋을것 같습니다.
findAnswerByWriter -> findByWriter

dadadamarine commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -64,14 +67,36 @@ public boolean isQuestionWriter(User loginUser, long id) {
-        // TODO 답변 추가 기능 구현
-        return null;
+    public Answer addAnswer(User loginUser, long questionId, AnswerDTO answerDTO) {
+        Question question = findById(questionId).get();
```
해당 id의 question이 없을경우에 대한 예외처리가 필요해 보이는 문장입니다.
혹시 answer생성자 안에서 예외처리를 하는건지 확인해 보았지만 보이지 않네요.

Bellroute replied:
get()안에 있지 않나요?

wijehyeon replied:
Optional 이라서 get()해오는 것이 자체가 불안한 상태인거같습니다

dadadamarine commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -64,14 +67,36 @@ public boolean isQuestionWriter(User loginUser, long id) {
+
+        question.addAnswer(answer);
+        questionRepository.save(question);
+
```
CascadeType에 대한 내용을 확인하고 적으신건지 궁금합니다.

현재 question에 answer에 대한 CascadeType가 적용되어있기때문에 에러가 나지 않더라도
가독성을 위해 answer를 save하는 내용도 같이 적어주는건 어떨까요.
(이 코드를 보는 사람마다 cascadeType를 각각 확인해볼거 같아서 효율적이지 않아보입니다.)

음.. 이부분은 사실 저도 부족하다고 생각해서 다른팀원들이 피드백을 더 해주었으면 좋겠습니다

Bellroute replied:
CascadeType에 대해 모르고 있었습니다. 이 부분 공부를 해보고 다시 고민해 반영하겠습니다

dadadamarine replied:
너무 학습할 양이많아지니  이부분은 jpa학습이후 반영하셔도 괜찮을것같습니다

hyukjin-lee replied:
반영 안할거 같으니까  지금 공부하세요

dadadamarine commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -64,14 +67,36 @@ public boolean isQuestionWriter(User loginUser, long id) {
+
+    public void deleteAnswer(User loginUser, long id) throws CannotDeleteException {
+        Answer answer = answerRepository
+                .findById(id).orElseThrow(NoSuchElementException::new);
```
위에 question을 탐색할때는 findById라는 qnaService의 메서드를 사용하셨고, 
answer를 탐색할때는 바로 answerRepository에서 접근도록 코드가 작성되어있습니다.
코드의 일관성을 위해 같은 방식으로 맞춰주는게 좋을것 같습니다.

+ 아래 findAnswerById가 구현되어있었네요

dadadamarine commented on src/main/java/codesquad/service/QnaService.java:
```diff
@@ -64,14 +67,36 @@ public boolean isQuestionWriter(User loginUser, long id) {
+        }
+
+        answer.setContents(answerDTO.getContents());
+
```
Answer create에서는 생성자로 바로 DTO를 넘긴반면
update에서는 content값을빼서 전송하네요.
일관성이 필요해 보입니다.

dadadamarine commented on src/main/java/codesquad/web/ApiAnswerController.java:
```diff
@@ -0,0 +1,46 @@
+    @DeleteMapping("/{id}")
+    public void delete(@LoginUser User loginUser, @PathVariable long questionId, @PathVariable long id) throws CannotDeleteException {
+        qnaService.deleteAnswer(loginUser, id);
+    }
```
음 여기도몰라서 여쭤봅니다. 컨트롤러에서 아무값도 리턴하지 않을시에 클라이언트쪽에 어떤 상태코드와 body로 전송되나요?

Bellroute replied:
확인해보니 비어있네요. 테스트는 상태코드가 ok로 통과가 됬었습니다. delete가 되면 프론트에서 디비를 거치지 않고 삭제된다고 생각해서 void로 구현했는데 생각을 잘못하고 있었습니다. 수정 후 리뷰 반영하겠습니다

dadadamarine commented on src/main/java/codesquad/web/ApiQuestionController.java:
```diff
@@ -0,0 +1,40 @@
+
+    @GetMapping("/{id}")
+    public Question show(@PathVariable long id) {
+        return qnaService.findById(id).orElseThrow(NoSuchElementException::new);
```
당연히 서비스에 있어야할 데이터 탐색에 대한 예외처리가 컨트롤러단까지 올라와 있는 느낌을 받습니다.

dadadamarine reviewed (commented):
제가 너무 깐깐하게 군거같아서 죄송합니다.. 
파이팅

wijehyeon commented on src/main/java/codesquad/domain/Question.java:
```diff
@@ -112,4 +113,13 @@ public void delete(User loginUser) throws CannotDeleteException {
         this.deleted = true;
     }
+
+    public boolean equalsTitleAndContents(Question target) {
```
boolean 타입의 반환 메소드 네이밍은
조동사~ 로 시작하면 좋겠네요
is~ has~

Bellroute replied:
자바에서 제공하는 비교 메소드도 equals()로 사용되고 있어서 is를 사용하지 않았는데 바꾸는 것이 좋을까요?

wijehyeon replied:
전 바꾸는게 좋다고 생각합니다

wijehyeon reviewed (commented):
늦게왔더니 다들 리뷰를 선수쳐갔군요

Bellroute committed:
refactor: 리뷰반영
1. 도메인 레이어에서 user 조회
2. 테스트만을 위한 코드 지양
3. 레이어 간 일관성 있는 데이터 전달 코드
4. ApiAnswerController delete() 리턴값 변경
5. 중복 String 상수화

wijehyeon commented on src/test/java/codesquad/web/ApiAnswerAcceptanceTest.java:
```diff
@@ -0,0 +1,92 @@
+        assertThat(updateAnswer.equals(response.getBody())).isTrue();
+    }
+
+    private HttpEntity createHttpEntity(Object body) {
```
AcceptanceTest에 구현해두면
더 테스트할 게 많아질 때 확장성을 갖게될 지도

wijehyeon commented on src/test/java/codesquad/web/ApiQuestionAcceptanceTest.java:
```diff
@@ -0,0 +1,61 @@
+        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.FORBIDDEN);
+    }
+
+    private HttpEntity createHttpEntity(Object body) {
```
이 메소드 부분의 코드가 중복되네요 상위 클래스에 구현하는게 좋겠어요

wijehyeon reviewed (commented):
👍
